frontend/javascripts/oxalis/model/sagas/volume/volume_interpolation_saga.ts

```diff
@@ -0,0 +1,257 @@
+import {
+  enforceActiveVolumeTracing,
+  getActiveSegmentationTracingLayer,
+  isVolumeAnnotationDisallowedForZoom,
+} from "oxalis/model/accessors/volumetracing_accessor";
+import DataLayer from "oxalis/model/data_layer";
+import Dimensions from "oxalis/model/dimensions";
+import type { Saga } from "oxalis/model/sagas/effect-generators";
+import { select } from "oxalis/model/sagas/effect-generators";
+import VolumeLayer, { VoxelBuffer2D } from "oxalis/model/volumetracing/volumelayer";
+import { call } from "typed-redux-saga";
+import { createVolumeLayer, getBoundingBoxForViewport, labelWithVoxelBuffer2D } from "./helpers";
+
+const isEqual = cwise({
+  args: ["array", "scalar"],
+  body: function body(a: number, b: number) {
+    a = a === b ? 1 : 0;
+  },
+});
+
+const mul = cwise({
+  args: ["array", "scalar"],
+  body: function body(a: number, b: number) {
+    a = a * b;
+  },
+});
+
+const avg = cwise({
+  args: ["array", "array"],
+  body: function body(a: number, b: number) {
+    a = (a + b) / 2;
+  },
+});
+
+const absMax = cwise({
+  args: ["array", "array"],
+  body: function body(a: number, b: number) {
+    a = Math.abs(a) > Math.abs(b) ? a : b;
+  },
+});
+
+const assign = cwise({
+  args: ["array", "array"],
+  body: function body(a: number, b: number) {
+    a = b;
+  },
+});
+
+function copy(Constructor: Float32ArrayConstructor, arr: ndarray.NdArray): ndarray.NdArray {
+  const newArr = ndarray(new Constructor(arr.size), arr.shape, arr.stride);
+  assign(newArr, arr);
+  return newArr;
+}
+
+/*
+ * Computes a signed distance transform for an input nd array.
+ */
+function signedDist(arr: ndarray.NdArray) {
+  // Copy the input twice to avoid mutating it
+  arr = copy(Float32Array, arr);
+  const negatedArr = copy(Float32Array, arr);
+
+  // Normal distance transform for arr
+  distanceTransform(arr);
+
+  // Invert negatedArr (1 to 0 and 0 to 1)
+  isEqual(negatedArr, 0);
+  distanceTransform(negatedArr);
+  // Negate the distances
+  mul(negatedArr, -1);
+
+  // Create a combined array which contains positive
+  // distances for voxels outside of the labeled area
+  // and negative distances for voxels inside the labeled
+  // area.
+  absMax(arr, negatedArr);
+  return arr;
+}
+
+export default function* maybeInterpolateSegmentationLayer(
+  layer: VolumeLayer,
+  isDrawing: boolean,
+  activeTool: AnnotationTool,
+): Saga<void> {
+  const allowUpdate = yield* select((state) => state.tracing.restrictions.allowUpdate);
+  if (!allowUpdate) return;
+  const activeViewport = yield* select((state) => state.viewModeData.plane.activeViewport);
+
+  if (!ToolsWithInterpolationCapabilities.includes(activeTool) || !isDrawing) {
+    return;
+  }
+
+  const isVolumeInterpolationEnabled = yield* select(
+    (state) => state.userConfiguration.isVolumeInterpolationEnabled,
+  );
+
+  if (!isVolumeInterpolationEnabled) {
+    return;
+  }
+
+  if (activeViewport !== "PLANE_XY") {
+    // Interpolation is only done/supported in XY
```

Comment:
One open question is how this could be communicated to the user. Since you're showing a toast if the resolution is too low, the same could be done here, maybe